stdlib/public/core/NormalizedCodeUnitIterator.swift

```diff
@@ -10,4 +10,513 @@
+  }
+  return numericCast(count)
+}
+
+//
+// Pointer casting helpers
+//
+@inline(__always)
+private func _unsafeMutableBufferPointerCast<T, U>(
+  _ ptr: UnsafeMutablePointer<T>,
+  _ count: Int,
+  to: U.Type = U.self
+) -> UnsafeMutableBufferPointer<U> {
+  return UnsafeMutableBufferPointer(
+    start: UnsafeMutableRawPointer(ptr).assumingMemoryBound(to: U.self),
+    count: count
+  )
+}
+@inline(__always)
+private func _unsafeBufferPointerCast<T, U>(
+  _ ptr: UnsafePointer<T>,
+  _ count: Int,
+  to: U.Type = U.self
+) -> UnsafeBufferPointer<U> {
+  return UnsafeBufferPointer(
+    start: UnsafeRawPointer(ptr).assumingMemoryBound(to: U.self),
+    count: count
+  )
+}
+
+internal func _castOutputBuffer(
+  _ ptr: UnsafeMutablePointer<_Normalization._SegmentOutputBuffer>,
+  endingAt endIdx: Int = _Normalization._SegmentOutputBuffer.capacity
+) -> UnsafeMutableBufferPointer<UInt16> {
+  let bufPtr: UnsafeMutableBufferPointer<UInt16> =
+    _unsafeMutableBufferPointerCast(
+      ptr, _Normalization._SegmentOutputBuffer.capacity)
+  return UnsafeMutableBufferPointer<UInt16>(rebasing: bufPtr[..<endIdx])
+}
+internal func _castOutputBuffer(
+  _ ptr: UnsafePointer<_Normalization._SegmentOutputBuffer>,
+  endingAt endIdx: Int = _Normalization._SegmentOutputBuffer.capacity
+) -> UnsafeBufferPointer<UInt16> {
+  let bufPtr: UnsafeBufferPointer<UInt16> =
+    _unsafeBufferPointerCast(
+      ptr, _Normalization._SegmentOutputBuffer.capacity)
+  return UnsafeBufferPointer<UInt16>(rebasing: bufPtr[..<endIdx])
+}
+
+extension UnsafeBufferPointer where Element == UInt8 {
+  internal func isStartOfLastCodeUnit(_ index: Int) -> Bool {
+    let nextIndex = index + 1
+    return nextIndex >= self.count
+  }
+  
+  internal func hasNormalizationBoundary(after index: Int) -> Bool {
+    if isStartOfLastCodeUnit(index) {
+      return true
+    }
+
+    if _isContinuation(self[index+1]) {
+      return false
+    }
+
+    let nextCU = utf8Scalar(startingAt: index+1)
+    return _hasNormalizationBoundary(before: nextCU)
+  }
+  
+  internal func hasNormalizationBoundary(before index: Int) -> Bool {
+    if index == 0 || index == count {
+      return true
+    }
+    
+    assert(!_isContinuation(self[index]))
+    
+    let cu = utf8Scalar(startingAt: index)
+    return _hasNormalizationBoundary(before: cu)
+  }
+}
+
+extension _StringGuts {
+  internal func foreignHasNormalizationBoundary(
+    after index: String.Index
+  ) -> Bool {
+    if foreignIsStartOfLastCodeUnit(index) {
+      return true
+    }
+    
+    let nextCU = foreignErrorCorrectedUTF16CodeUnit(at: index._next())
+    return _hasNormalizationBoundary(before: nextCU)
+  }
+  
+  internal func foreignHasNormalizationBoundary(
+    before index: String.Index
+  ) -> Bool {
+    let offset = index.encodedOffset
+    if offset == 0 || offset == count {
+      return true
+    }
+    
+    let cu = foreignErrorCorrectedUTF16CodeUnit(at: index)
+    return _hasNormalizationBoundary(before: cu)
```

Comment:
Why did you decide to do a code unit? You have `_hasNormalizationBoundary(before:)` returning false for surrogates, is this intentional?